### PR TITLE
[Refactor](function) make all Datetime arithmetic operation overflow lead to exception in BE

### PR DESCRIPTION
**Pull request by zclllyybb**
### What problem does this PR solve?

Issue Number: close #xxx

Related PR: #xxx

Problem Summary:

1. totally refactored `FunctionDateOrDateTimeComputation`. removed some unnecessary function and template. simplified some template calculations.
2. All Datetime arithmetic operation overflow will lead to exception now. before for nullable input it will get `NULL` result
see: 
```sql
mysql> select date_add('5000-10-10', interval 10000 year);
+------------------------------------------------+
| years_add(cast('5000-10-10' as DATEV2), 10000) |
+------------------------------------------------+
| NULL                                           |
+------------------------------------------------+
1 row in set (0.10 sec)
```
now:
```sql
ERROR 1105 (HY000): errCode = 2, detailMessage = (xxx)[E-218][E-218] Operation years_add of 5000-10-10, 10000 out of range
```

### Release note

All Datetime arithmetic operation overflow will lead to exception now.

### Check List (For Author)

- Test 
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [x] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [x] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason 

- Behavior changed:
    - [ ] No.
    - [x] Yes. 

- Does this need documentation?
    - [ ] No.
    - [x] Yes. https://github.com/apache/doris-website/pull/2176

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Review by Mryange — APPROVED** [review]

**Review by HappenLee — CHANGES_REQUESTED** [review]
PLEASE ADD DETAIL OF THE CHANGE

File: be/src/vec/functions/function_date_or_datetime_computation.h [hunk 1]
```diff
@@ -318,381 +278,114 @@ struct TimeDiffImpl {
...
-        size_t size = vec_from0.size();
-        vec_to.resize(size);
-
-        bool invalid = true;
-        for (size_t i = 0; i < size; ++i) {
             vec_to[i] = Transform::execute(vec_from0[i], vec_from1[i], invalid);
 
             if (UNLIKELY(invalid)) {
-                throw Exception(ErrorCode::OUT_OF_BOUND, "Operation {} {} {} out of range",
+                throw Exception(ErrorCode::OUT_OF_BOUND, "Operation {} of {} {} out of range",
```
Comment (HappenLee) [inline be/src/vec/functions/function_date_or_datetime_computation.h:334]: format the `vec_from0[i]`

**Comment by zclllyybb** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**